test/acl.test.js

```diff
@@ -472,3 +475,138 @@ describe('access check', function() {
       });
   });
 });
+
+describe('Authorized roles propagated in RemotingContext', function() {
+  var app, request, accessToken;
+  var models = {};
+
+  beforeEach(setupAppAndRequest);
+
+  it('contains all authorized roles for a principal if query is allowed', function() {
+    return createACLs('MyTestModel', [
+      {permission: ACL.ALLOW, principalId: '$everyone'},
+      {permission: ACL.ALLOW, principalId: '$authenticated'},
+      {permission: ACL.ALLOW, principalId: 'myRole'},
+    ])
+    .then(function() {
+      return request.get('/MyTestModels')
+        .set('X-Access-Token', accessToken.id)
+        .expect(200);
+    })
+    .then(function(res) {
+      return models.MyTestModel.onBeforeRemote;
+    })
+    .then(function(ctx) {
+      expect(ctx.args.options.authorizedRoles).to.eql(
+        ['$everyone', '$authenticated', 'myRole']
+      );
+    });
+  });
+
+  it('does not contain any denied role even if query is allowed', function() {
+    return createACLs('MyTestModel', [
+      {permission: ACL.ALLOW, principalId: '$everyone'},
+      {permission: ACL.DENY,  principalId: '$authenticated'},
+      {permission: ACL.ALLOW, principalId: 'myRole'},
+    ])
+    .then(function() {
+      return request.get('/MyTestModels')
+        .set('X-Access-Token', accessToken.id)
+        .expect(200);
+    })
+    .then(function(res) {
+      return models.MyTestModel.onBeforeRemote;
+    })
+    .then(function(ctx) {
+      expect(ctx.args.options.authorizedRoles).to.eql(
+        ['$everyone', 'myRole']
+      );
+    });
+  });
+
+  it('honors default permission setting', function() {
+    // set MyTestModel default permission as DENY
+    models.MyTestModel.settings.defaultPermission = ACL.DENY;
+
+    return createACLs('MyTestModel', [
+      {permission: ACL.DEFAULT, principalId: '$everyone'},
+      {permission: ACL.DENY,    principalId: '$authenticated'},
+      {permission: ACL.ALLOW,   principalId: 'myRole'},
+    ])
+    .then(function() {
+      return request.get('/MyTestModels')
+        .set('X-Access-Token', accessToken.id)
+        .expect(200);
+    })
+    .then(function(res) {
+      return models.MyTestModel.onBeforeRemote;
+    })
+    .then(function(ctx) {
+      expect(ctx.args.options.authorizedRoles).to.eql(
+        // '$everyone' is not expected as default permission is DENY
+        ['myRole']
+      );
+    });
+  });
+
+  // helpers
+  function setupAppAndRequest() {
+    app = loopback({localRegistry: true, loadBuiltinModels: true});
+    app.use(loopback.rest());
+    app.set('remoting', {errorHandler: {debug: true, log: true}});
+    app.enableAuth();
+    app.dataSource('db', {connector: 'memory'});
+    request = supertest(app);
+
+    // attaching base models
+    ['User', 'Role', 'RoleMapping', 'ACL', 'AccessToken'].map(setupModel);
```

Comment:
- `app.enableAuth({ dataSource: 'db' })` automatically attaches all required models to the given datasource.
- any app model can be accessed via `app.models`, so it should be enough to set `var models = app.models;`

```js
app.enableAuthentication({dataSource: 'db'});
models = app.models;
```

This way, `setupModel` shouldn't be needed at all.